quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -416,66 +311,168 @@ impl IngestController {
                 };
                 get_or_create_open_shards_successes.push(get_or_create_open_shards_success);
             } else {
-                // TODO: Find leaders in batches.
-                // TODO: Round-robin leader-follower pairs or choose according to load.
-                let (leader_id, follower_id) = self
-                    .find_leader_and_follower(&mut unavailable_leaders, progress)
-                    .await
-                    .ok_or_else(|| {
-                        ControlPlaneError::Unavailable("no ingester available".to_string())
-                    })?;
                 let shard_id = ShardId::from(Ulid::new());
                 let open_shards_subrequest = metastore::OpenShardsSubrequest {
                     subrequest_id: get_open_shards_subrequest.subrequest_id,
                     index_uid: index_uid.into(),
                     source_id: get_open_shards_subrequest.source_id,
                     shard_id: Some(shard_id),
-                    leader_id: leader_id.into(),
-                    follower_id: follower_id.map(|follower_id| follower_id.into()),
+                    // These attributes will be overwritten in the next stage.
+                    leader_id: "".to_string(),
+                    follower_id: None,
                 };
                 open_shards_subrequests.push(open_shards_subrequest);
             }
         }
         if !open_shards_subrequests.is_empty() {
-            let open_shards_request = metastore::OpenShardsRequest {
-                subrequests: open_shards_subrequests,
-            };
-            let open_shards_response = progress
-                .protect_future(self.metastore.open_shards(open_shards_request))
-                .await?;
-
-            // TODO: Handle failures.
-            let _ = self.init_shards(&open_shards_response, progress).await;
-
-            for open_shards_subresponse in open_shards_response.subresponses {
-                let index_uid: IndexUid = open_shards_subresponse.index_uid().clone();
-                let source_id = open_shards_subresponse.source_id.clone();
-                model.insert_shards(
-                    &index_uid,
-                    &source_id,
-                    open_shards_subresponse.opened_shards,
-                );
-                if let Some(open_shard_entries) =
-                    model.find_open_shards(&index_uid, &source_id, &unavailable_leaders)
+            if let Some(leader_follower_pairs) =
```

Comment:
This function has become quite lengthy and messy. I will refactor it in a follow-up PR, most likely with a workbench, since many steps are involved in opening a new shard.